File: Current/Product/Production/Nant/Common.Tasks/SaveProperties.vb
Option Explicit On
Option Strict On

Imports NAnt.Core
Imports NAnt.Core.Attributes

Imports System.IO

<TaskName("saveproperties")> _
Public Class SaveProperties
    Inherits Task

    Public Enum FileFormat
        Include
        CommandLine
    End Enum

    Private _Format As FileFormat
    Private _File As FileInfo
    Private _PropertyList As SaveProperty()
    Private _Append As Boolean = False
    Private _ProjectName As String = "Bogas"

    <TaskAttribute("projectname", required:=False)> _
    Public Property ProjectName() As String
        Get
            Return _ProjectName
        End Get
        Set(ByVal value As String)
            _ProjectName = value
        End Set
    End Property

    <TaskAttribute("append", Required:=False), BooleanValidator()> _
    Public Property Append() As Boolean
        Get
            Return _Append
        End Get
        Set(ByVal value As Boolean)
            _Append = value
        End Set
    End Property

    <BuildElementArray("property", Required:=True)> _
    Public Property PropertyList() As SaveProperty()
        Get
            Return _PropertyList
        End Get
        Set(ByVal value As SaveProperty())
            _PropertyList = value
        End Set
    End Property

    <TaskAttribute("file", Required:=True)> _
    Public Property File() As FileInfo
        Get
            Return _File
        End Get
        Set(ByVal value As FileInfo)
            _File = value
        End Set
    End Property

    <TaskAttribute("format", Required:=True)> _
    Public Property Format() As FileFormat
        Get
            Return _Format
        End Get
        Set(ByVal value As FileFormat)
            _Format = value
        End Set
    End Property

    Protected Overrides Sub ExecuteTask()
        Select Case Me.Format
            Case FileFormat.CommandLine
                Me.CreateCommandLineFile()
            Case FileFormat.Include
                Me.CreateIncludeFile()
            Case Else
                Throw New BuildException(String.Format("Format {0} is not supported.", Me.Format.ToString))
        End Select
    End Sub

    Public Sub CreateIncludeFile()
        Dim Writer As StreamWriter
        Try
            Writer = Me.GetWriter
            Writer.WriteLine("<?xml version='1.0' encoding='utf-8' ?>")
            Writer.WriteLine(String.Format("<project xmlns='http://nant.sf.net/schemas/nant.xsd'  name='{0}'>", Me.ProjectName))
            For Each [Property] As SaveProperty In Me.PropertyList
                Writer.WriteLine(String.Format("<property name='{0}' value='{1}' />", [Property].PropertyName, [Property].PropertyValue))
            Next
            Writer.WriteLine("</project>")
        Finally
            If Not Writer Is Nothing Then
                Writer.Close()
            End If
        End Try
    End Sub

    Public Sub CreateCommandLineFile()
        Dim Writer As StreamWriter
        Try
            Writer = Me.GetWriter
            For Each [Property] As SaveProperty In Me.PropertyList
                Writer.WriteLine(String.Format("-D:{0}=""{1}""", [Property].PropertyName, [Property].PropertyValue))
            Next
        Finally
            If Not Writer Is Nothing Then
                Writer.Close()
            End If
        End Try
    End Sub

    Private Function GetWriter() As StreamWriter
        If Me.Append Then
            Return Me.File.AppendText
        Else
            Return Me.File.CreateText
        End If
    End Function

End Class

File: Product/Production/Nant/CIFactory.NAnt.Tasks/Tasks/SaveProperties.cs
using System;
using System.IO;
using NAnt.Core;
using NAnt.Core.Attributes;
using CIFactory.NAnt.Types;

namespace CIFactory.NAnt.Tasks
{
    [TaskName("saveproperties")]
    public class SaveProperties : Task
    {
        #region Enums

        public enum FileFormat
        {
            Include,
            CommandLine
        }

        #endregion

        #region Fields

        private bool _Append = false;

        private FileInfo _File;

        private FileFormat _Format;

        private string _ProjectName;

        private SaveProperty[] _PropertyList;

        #endregion

        #region Properties

        [TaskAttribute("append", Required = false), BooleanValidator()]
        public bool Append
        {
            get { return _Append; }
            set { _Append = value; }
        }

        [TaskAttribute("file", Required = true)]
        public FileInfo File
        {
            get { return _File; }
            set { _File = value; }
        }

        [TaskAttribute("format", Required = true)]
        public FileFormat Format
        {
            get { return _Format; }
            set { _Format = value; }
        }

        [TaskAttribute("projectname", Required = false)]
        public string ProjectName
        {
            get { return _ProjectName; }
            set { _ProjectName = value; }
        }

        [BuildElementArray("property", Required = true)]
        public SaveProperty[] PropertyList
        {
            get { return _PropertyList; }
            set { _PropertyList = value; }
        }

        #endregion

        #region Public Methods

        public void CreateCommandLineFile()
        {
            StreamWriter Writer = null;
            try
            {
                Writer = this.GetWriter();
                foreach (SaveProperty Property in this.PropertyList)
                {
                    Writer.WriteLine(string.Format("-D:{0}=\"{1}\"", Property.PropertyName, Property.PropertyValue));
                }
            }
            finally
            {
                if (Writer != null)
                {
                    Writer.Close();
                }
            }
        }

        public void CreateIncludeFile()
        {
            StreamWriter Writer = null;
            try
            {
                Writer = this.GetWriter();
                Writer.WriteLine("<?xml version='1.0' encoding='utf-8' ?>");
                if (string.IsNullOrEmpty(this.ProjectName))
                    Writer.WriteLine("<project xmlns='http://nant.sf.net/schemas/nant.xsd'>");
                else
                    Writer.WriteLine(string.Format("<project xmlns='http://nant.sf.net/schemas/nant.xsd'  name='{0}'>", this.ProjectName));
                foreach (SaveProperty Property in this.PropertyList)
                {
                    Writer.WriteLine(string.Format("<property name='{0}' value='{1}' />", Property.PropertyName, Property.PropertyValue));
                }
                Writer.WriteLine("</project>");
            }
            finally
            {
                if (Writer != null)
                {
                    Writer.Close();
                }
            }
        }

        #endregion

        #region Protected Methods

        protected override void ExecuteTask()
        {
            switch (this.Format)
            {
                case FileFormat.CommandLine:
                    this.CreateCommandLineFile();
                    break;
                case FileFormat.Include:
                    this.CreateIncludeFile();
                    break;
                default:
                    throw new BuildException(string.Format("Format {0} is not supported.", this.Format.ToString()));
            }
        }

        #endregion

        #region Private Methods

        private StreamWriter GetWriter()
        {
            if (this.Append)
            {
                return this.File.AppendText();
            }
            else
            {
                return this.File.CreateText();
            }
        }

        #endregion

    }

}